core/src/main/scala/cats/Monad.scala

```diff
@@ -14,4 +14,90 @@ import simulacrum.typeclass
 @typeclass trait Monad[F[_]] extends FlatMap[F] with Applicative[F] {
   override def map[A, B](fa: F[A])(f: A => B): F[B] =
     flatMap(fa)(a => pure(f(a)))
+
+  /**
+   * Execute an action repeatedly as long as the given `Boolean` expression
+   * returns `true`. The condition is evaluated before the loop body.
+   * Collects the results into an arbitrary `Alternative` value, such as a `Vector`.
+   * This implementation uses append on each evaluation result,
+   * so avoid data structures with non-constant append performance, e.g. `List`.
+  */
+  def whileM[G[_], A](p: F[Boolean])(body: => F[A])(implicit G: Alternative[G]): F[G[A]] = {
+    val b = Eval.later(body)
+    tailRecM[G[A], G[A]](G.empty)(xs => ifM(p)(
+      ifTrue = {
+        map(b.value) { bv =>
+          Left(G.combineK(xs, G.pure(bv)))
+        }
+      },
+      ifFalse = pure(xs.asRight[G[A]])
+    ))
+  }
+
+  /**
+   * Execute an action repeatedly as long as the given `Boolean` expression
+   * returns `true`. The condition is evaluated before the loop body.
+   * Discards results.
+   */
+  def whileM_[A](p: F[Boolean])(body: => F[A]): F[Unit] = {
+    val continue: Either[Unit, Unit] = Left(())
+    val stop: F[Either[Unit, Unit]] = pure(Right(()))
+    val b = Eval.later(body)
```

Comment:
That's fine, I was mostly asking because of the `Eval.later`. If we used `Eval.always` then you could use `Try` or `Future` by sneaking some effect into the call-by-name param, but I'm not a fan of that.

If we want to stick with `Eval.later` over lazy val, I'd like us to document the reason. The only reason I can see is that it allows the thunk in the call-by-name to be GC'ed after we run, which can be nice. We should comment why because someone will forget and revert since Later is only more work (it uses lazy val + some work).